flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -47,23 +50,33 @@ public class SharedStateRegistryImpl implements SharedStateRegistry {
     /** All registered state objects by an artificial key */
     private final Map<SharedStateRegistryKey, SharedStateEntry> registeredStates;
 
+    /** All registered checkpoints */
+    private final Set<CompletedCheckpoint> registeredCheckpoints;
```

Comment:
I realized that this field is now (partially) duplicating the `DefaultCompletedCheckpointStore.completedCheckpoints`.

Do you think it could/should be refactored so that `CompletedCheckpointStore` "transfers the ownership" of subsumed checkpoints to `SharedStateRegistry`? E.g. by passing them to `unregisterUnusedState`.

Another issue: `CompletedCheckpoint` doesn't have `equals/hashCode` overriden, so it's better to use `Map<Long, CompletedCheckpoint>`.